research/02-repl.tex

```diff
@@ -1,5 +1,172 @@
+    parses this input and creates an abstract syntax tree (AST).
+  \item The AST created in the previous step is then passed on to the
+    \texttt{eval} function, which evaluates it.
+  \item The result yielded by the previous function is then printed out to the
+    user by the \texttt{print} function.
+  \item After having printed the result, the environment needs to \texttt{loop}
+    back to the read state.
+\end{enumerate}
+
+Assuming the individual functions listed previously exist, a REPL can be created
+in a single line of code simply by combining the functions:
+
+\begin{lstlisting}
+(loop (print (eval (read))))
+\end{lstlisting}
+
+Part of the semantic differences between the different names for an interactive
+programming environment is Lisp's homoiconicity. In Lisp, the structure of
+expressions is represented directly in a data structure, resulting in the
+ability to infer a program's or data object's state simply by reading its
+textual representation. In Lisp REPLs, therefore, arbitrary data objects yielded
+from a previous expression can be used as input to the next expression.
+Homoiconicity is an unusual feature in modern programming languages not
+belonging to the Lisp family, however, so interactive environments of these
+programming languages often require additional hoops to read and evaluate
+expressions. As said earlier, this difference is mostly semantic and therefore
+this report sticks with the name \textit{REPL}.
+
+REPLs offer several advantages that make them an often requested feature.
+Because of their nature, REPLs provide the ability to program interactively.
+%Interactive programming is a programming paradigm in which parts of a program
+%are written while it is already active~\cite{NEEDED}.
+Programming interactively has multiple applications.
+
+% TODO: mention REPL Driven Development (RDD)?
+When creating software solutions for an as of yet not well understood domain,
+often it is not clear which data structures and algorithms are required. In such
+cases, REPLs offer the ability to interactively develop and debug software
+without having to apply the (oftentimes much slower) edit-compile-run-debug
+development style. This kind of programming is called exploratory
+programming~\cite{NEEDED}. Related to this kind of exploration, REPLs also
+provide a means for fast prototyping and bottom up programming~\cite{NEEDED}.
+
+The explorative and interactive features of a REPL also make it an excellent
+tool for programmers to learn a new programming language. REPLs are also
+combined with what is called literate programming to offer notebooks or language
+playgrounds, as discussed in~\cref{ssec:literate-programming}.
+
+Depending on one's definition of a REPL, a command line shell such as
+GNU Bash~\cite{NEEDED} can also be considered a REPL. In this case, one
+additional functionality of a REPL is to provide an interface to an operating
+system's functions.
```

Comment:
This paragraph feels out of place, the preceding paragraphs all talk about the advantages. Suddenly it is about an alternate definition/use case or metaphor of a REPL.